File: 2015/11/11.cpp
#include <iostream>
#include <string>
#include <unordered_map>

const std::string valid_alphabet ("abcdefghjkmnpqrstuvwxyz");
const std::string invalid_starts ("ghijklmnoyz");

bool check_pairs(std::string& s_in) {
    char prev = s_in.at(0);
    int found = 0;
    std::string used {""};
    for ( std::string::iterator it=s_in.begin() + 1; it!=s_in.end(); ++it)
        if (*it == prev) {
            found += 1;
            prev = 'A';
            used.push_back(*it);
        }
        else if (used.find(*it) != std::string::npos) {
            continue;
        }
        else {
            prev = *it;
        }
    if (found == 2) {
        return true;
    };
    return false;
}

bool check_straight(std::string& s_in) {
    char c {'A'};
    for (int i = 0; i<s_in.length(); i++) {
        c = s_in.at(i);
        if (invalid_starts.find(c) != std::string::npos) {
           continue; 
        }
        else if (i > s_in.length() - 3) {
            return false;
        }
        else if ((s_in.at(i+1) == valid_alphabet.at(valid_alphabet.find(s_in.at(i)) + 1)) && (s_in.at(i+2) == valid_alphabet.at(valid_alphabet.find(s_in.at(i)) + 2))) {
            return true;            
        }
    }
    return false;
}

void increment_password(std::string& s_in) {
    bool done = false;
    size_t c_idx = s_in.length() - 1;
    size_t s = 1;
    int c_inc = 0;
    while (!done) {
        c_inc = valid_alphabet.find(s_in.at(c_idx)) + 1;
        if (c_inc < valid_alphabet.length()) {
            done = true;
            s_in.replace(c_idx, s, &valid_alphabet.at(c_inc), 1);
        }
        else {
            s_in.replace(c_idx, s, &valid_alphabet.at(0), 1);
            c_idx -= 1;
        }
    }
}

int main() {
    std::string input = "hepxcrrq"; // Part I
    // std::string input = "hepxxyzz"; // Part II

    bool found_new = false;
    bool cond1 = false;
    bool cond2 = false;

    while (!found_new) {
        increment_password(input);
        cond1 = check_pairs(input);
        cond2 = check_straight(input);
        if (cond1 && cond2) {
            found_new = true;
        }
    }
    std::cout << input;

}

File: 2015/11/11.py
input = 'hepxcrrq' # Part I
# input = 'hepxxyzz' # Part II

valid_alphabet = 'abcdefghjkmnpqrstuvwxyz'
valid_alphabet_ind = {c:i for (i, c) in enumerate(valid_alphabet)}

def check_pairs(s_in):
  # O(n) method for checking there are two
  # non-overlapping pairs of repeated chars
  # it's unclear whether the pairs two non-overlapping
  # pairs can be the same letter. here we assume no.
  prev = s_in[0]
  found = 0
  used = []
  for c in s_in[1:]:
    if c == prev:
      found += 1
      prev = None
      used.append(c)
    elif c in used:
      pass
    else:
      prev = c
  if found == 2:
    return True
  return False

def check_straight(s_in):
  # O(n) method for checking if there's a three letter
  # straight
  for i, c in enumerate(s_in):
    if c in 'ghijklmnoyz':
      pass
    elif i > len(s_in) - 3: 
      return False
    elif (s_in[i+1] == valid_alphabet[valid_alphabet_ind[s_in[i]] + 1]) and (s_in[i+2] == valid_alphabet[valid_alphabet_ind[s_in[i]] + 2]): 
      return True
  

def increment_password(s_in):
  s_in = list(s_in)
  done = False
  c_idx = len(s_in) - 1
  while not done:
    c_inc = valid_alphabet_ind[s_in[c_idx]] + 1
    if c_inc < len(valid_alphabet):
      done = True
      s_in[c_idx] = valid_alphabet[c_inc]
    else:
      s_in[c_idx] = valid_alphabet[0]
      c_idx -= 1 
  return ''.join(s_in)

found_new = False
while not found_new:
  input = increment_password(input)
  if check_straight(input) and check_pairs(input):
    found_new = True
print(input)
